resources/assets/lib/osu-common.ts

```diff
@@ -0,0 +1,372 @@
+
+    if (newUrl === currentUrl) {
+      return;
+    }
+
+    history.replaceState(history.state, '', newUrl);
+  },
+
+  storeJson: (id: string, object: Record<string, unknown>) => {
+    const json = JSON.stringify(object);
+    let element = document.getElementById(id) as (HTMLScriptElement | null);
+
+    if (element == null) {
+      element = document.createElement('script');
+      element.id = id;
+      element.type = 'application/json';
+      document.body.appendChild(element);
+    }
+
+    element.text = json;
+  },
+
+  timeago: (time = '') => {
+    const el = document.createElement('time');
+
+    el.classList.add('js-timeago');
+    el.setAttribute('datetime', time);
+    el.textContent = time;
+
+    return el.outerHTML;
+  },
+
+  trans: (key: string, replacements = {}, locale?: string) => {
+    if (osu.transExists(key, locale)) {
+      locale = fallbackLocale;
+    }
+
+    return Lang.get(key, replacements, locale);
+  },
+
+  transArray: (array: any[], key = 'common.array_and') => {
+    switch (array.length) {
+      case 0:
+        return '';
+      case 1:
+        return String(array[0]);
+      case 2:
+        return array.join(osu.trans(`${key}.two_words_connector`));
+      default:
+        return `${array.slice(0, -1).join(osu.trans(`${key}.words_connector`))}${osu.trans(`${key}.last_word_connector`)}${String(array[array.length - 1])}`;
+    }
+  },
+
+  transChoice: (key: string, count: number, replacements: Record<string, unknown> = {}, locale?: string): string => {
+    locale ??= currentLocale;
+    const isFallbackLocale = locale === fallbackLocale;
+
+    if (!isFallbackLocale && !osu.transExists(key, locale)) {
+      return osu.transChoice(key, count, replacements, fallbackLocale);
+    }
+
+    replacements.count_delimited = osu.formatNumber(count, null, null, locale);
+    const translated = Lang.choice(key, count, replacements, locale);
+
+    if (!isFallbackLocale && translated == null) {
+      // added by Lang.choice
+      delete replacements.count;
+
+      return osu.transChoice(key, count, replacements, fallbackLocale);
+    }
+
+    return translated;
+  },
+
+  // Handles case where crowdin fills in untranslated key with empty string.
+  transExists: (key: string, locale?: string) => {
+    const translated = Lang.get(key, null, locale);
+
+    return osu.present(translated) && translated !== key;
+  },
+
+  updateQueryString: (url: string | null, params: { [key: string]: string | null | undefined }) => {
+    const docUrl = getCurrentUrl();
+    const urlObj = new URL(url ?? docUrl.href, docUrl.origin);
+
+    // FIXME: not sure about this conversion `for own` from coffeescript
+    for (const key in params) {
+      if (Object.prototype.hasOwnProperty.call(params, key)) {
+        const value = params[key];
+
+        if (value != null) {
+          urlObj.searchParams.set(key, value);
+        } else {
+          urlObj.searchParams.delete(key);
+        }
+      }
+    }
```

Comment:
Original coffeescript is 

```coffeescript
    for own key, value of params
      if value?
        urlObj.searchParams.set(key, value)
      else
        urlObj.searchParams.delete(key)
```

I'm not sure enough about conversion `for own` to plain javascript.